docs/adr/adr-11-blocksync-overhaul.md

```diff
@@ -0,0 +1,329 @@
+to truncate the CAR file to get strictly the first quadrant out of it without NMT proofs, so serialization must be
+different from the utility to support that.
+
+NOTE2: Alternatively to `WriteEDS`, and `EDSReader` could be introduced to make EDS-to-stream handling more idiomatic
+and efficient in some cases, with the cost of more complex implementation.
+
+```go
+// WriteEDS writes whole EDS into given io.Writer as CARv1 file.
+// All its shares and recomputed NMT proofs.
+func WriteEDS(context.Context, *rsmt2d.ExtendedDataSquare, io.Writer) error
+```
+
+###### `share.ReadEDS`
+
+To read EDS out of stream/file `ReadEDS` is introduced. Internally, it
+
+- Imports EDS with an empty pre-allocated slice. NOTE: Size can be taken from DataRoot
+- Wraps given io.Reader with [`BlockReader`](https://github.com/ipld/go-car/blob/master/v2/block_reader.go#L16)
+- Reads out blocks one by one and fills up the EDS quadrant via `EDS.SetCell`
+- Recomputes and validates via `EDS.Repair`
+
+```go
+// ReadEDS reads EDS quadrant(1/4) from io.Reader CAR file.
+//
+// It expects strictly first EDS quadrant(top left).
+// Returned EDS is guaranteed to be full and valid against DataRoot, otherwise ReadEDS errors.
+func ReadEDS(context.Context, io.Reader, DataRoot) (*rsmt2d.ExtendedDataSquare, error)
+```
+
+##### `share.EDSStore`
+
+To manage every EDS on the disk FNs/BNs keep `EDSStore` type is introduced in `share` pkg. Each EDS together with its
+Merkle Proofs serializes into CARv1 file. All the serialized CARv1 file blobs are stored as OS FS files as DAGStore
+[Mounts](https://github.com/filecoin-project/dagstore/blob/master/mount/mount.go).
+
+The introduced `EDSStore` also maintains (via DAGStore) a top-level index enabling granular and efficient random access
+to every share and/or Merkle proof over every registered CARv1 file. The `EDSStore` provides a custom `Blockstore` interface
+implementation to achieve the access. The main use-case is randomized sampling over the whole chain of EDS block data
+and getting data by namespace.
+
+EDSStore constructor should instantiate
+
+```go
+type EDSStore struct {
+ basepath string 
+ dgstr dagstore.DAGStore
+ topIdx index.Inverted 
+ carIdx index.FullIndexRepo
+ mounts  *mount.Registry
+ ...
+}
+
+// NewEDSStore constructs EDStore over OS directory to store CARv1 files of EDSes and indices for them.
+// Datastore is used to keep inverted/top-level index.
+func NewEDSStore(basepath string, ds datastore.Batching) *EDSStore {
+ topIdx := index.NewInverted(datastore)
+ carIdx := index.NewFSRepo(basepath + "/index")
+ mounts := mount.NewRegistry()
+ return &EDSStore{
+  basepath: basepath,
+  dgst: dagstore.New(dagstore.Config{...}),
+  topIdx: index.NewInverted(datastore),
+  carIdx: index.NewFSRepo(basepath + "/index")
+  mounts: mounts,
+    }   
+}
+
+```
+
+##### `share.EDSStore.Put`
+
+To write an entire EDS `Put` method is introduced. Internally it
+
+- Opens a file under `storepath/DataRoot` path
```

Comment:
```suggestion
- Opens a file under `storepath/DataRoot` path using `OpenReadWrite` creating a CAR-level blockstore
```

Reply:
- It doesnt require us to use CARv2 (there is an option on it called `WriteAsCarV1`)
- It looks like it does lay down things in order, but I cannot tell if it will be fully compatible
- It isn't clear to me yet why either indexing timepoint would be better, so I trust your instincts there

But yes I think we can leave it out for now, not sure its relevant for the ADR

Reply:
It's not relevant, so resolving. WriteEDS now describes almost custom way to write the CARv1 while which `blockstore.OpenReadWrite` won't support, so resolved